juju/juju
# Report storage within FullStatus facade call.

Instead of `juju status --storage` making 3 seperate facade calls in addition to the FullStatus call, instead return the information in the FullStatus facade call.

This also implements server side filtering on unit/application for storage entities.

## QA steps

- Deploy multiple apps with storage.
- Check `juju status` without storage
- Check `juju status --storage`
- Check `juju status --storage` works against old controller (Client facade <= 6)
- Check `juju status --format yaml` (should have implicit --storage)
- Check `juju status --storage <filter>`

Example no filter:
```
juju status --storage
Model  Controller  Cloud/Region  Version  SLA          Timestamp
a      minikube    minikube      3.2.3    unsupported  16:13:46+10:00

App             Version  Status   Scale  Charm           Channel    Rev  Address         Exposed  Message
b                        waiting      1  postgresql-k8s  14/stable   73  10.110.113.233  no       installing agent
postgresql-k8s           waiting      1  postgresql-k8s  14/stable   73  10.104.229.135  no       installing agent

Unit               Workload  Agent  Address     Ports  Message
b/0*               blocked   idle   10.244.0.6         failed to create k8s resources
postgresql-k8s/0*  blocked   idle   10.244.0.7         failed to create k8s resources

Storage Unit      Storage ID  Type        Pool        Mountpoint                Size     Status    Message
b/0               pgdata/1    filesystem  kubernetes  /var/lib/postgresql/data  1.0 GiB  attached  Successfully provisioned volume pvc-f08518a7-20be-4184-bc50-e668baa7fa37
postgresql-k8s/0  pgdata/0    filesystem  kubernetes  /var/lib/postgresql/data  1.0 GiB  attached  Successfully provisioned volume pvc-48b7f0ea-b103-4d66-b21e-b69d107a072d
```

Example with filter:
```
juju status --storage b
Model  Controller  Cloud/Region  Version  SLA          Timestamp
a      minikube    minikube      3.2.3    unsupported  16:16:27+10:00

App  Version  Status   Scale  Charm           Channel    Rev  Address         Exposed  Message
b    14.7     waiting      1  postgresql-k8s  14/stable   73  10.110.113.233  no       installing agent

Unit  Workload  Agent  Address     Ports  Message
b/0*  blocked   idle   10.244.0.6         failed to create k8s resources

Storage Unit  Storage ID  Type        Pool        Mountpoint                Size     Status    Message
b/0           pgdata/1    filesystem  kubernetes  /var/lib/postgresql/data  1.0 GiB  attached  Successfully provisioned volume pvc-f08518a7-20be-4184-bc50-e668baa7fa37
```

## Documentation changes

N/A

## Bug reference

Partially mitigates https://bugs.launchpad.net/juju/+bug/2015371

JUJU-4538
